### PR TITLE
Feat/complex functions

### PR DESCRIPTION
**Pull request by jtlap**
complex functions as libc++

### CODE REVIEW

File: include/eve/module/complex/regular/complex.hpp [hunk 1]
```diff
@@ -11,3 +11,9 @@
 #include <eve/module/complex/regular/real.hpp>
 #include <eve/module/complex/regular/i.hpp>
 #include <eve/module/complex/regular/imag.hpp>
+#include <eve/module/complex/regular/exp_i.hpp>
+#include <eve/module/complex/regular/exp_ipi.hpp>
+#include <eve/module/complex/regular/polar.hpp>
+#include <eve/module/complex/regular/detail/acos.hpp>
```
Comment (jfalcou) [inline include/eve/module/complex/regular/complex.hpp:17]: I would prefer those 3 includes to live at the end of detail/math.hpp

File: include/eve/module/complex/regular/detail/acos.hpp [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+        }
+        not_treated = logical_notand(zone4, not_treated);
+        if (eve::any(not_treated))
+        {
+          rtype aa = eve::sqrt(eve::inc(sqr(y)));
+          r = if_else(not_treated, pio_2(as(x)), r);
+          i = if_else(not_treated, half(as(x))*eve::log1p(2*y*(y+aa)), i);
+        }
+      }
+      if (eve::any(eve::is_not_finite(a0))) // this is perhaps for pedantic specialization ?
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/acos.hpp:122]: this needs to live in a pedantic function

File: include/eve/module/complex/regular/detail/arithmetic.hpp [hunk 3]
```diff
@@ -33,21 +42,55 @@ namespace eve::detail
...
+    return sum_of_prod(zr, zr, zi, zi);
+  }
+
+
+  //==============================================================================================
+  //  Binary functions
+  //==============================================================================================
+
+  template<typename Z>
+  EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::add_, Z const& x, Z const& y) noexcept
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/arithmetic.hpp:64]: this is unneeded as add use + already

File: include/eve/module/complex/regular/detail/arithmetic.hpp [hunk 4]
```diff
@@ -33,21 +42,55 @@ namespace eve::detail
...
+  {
+    return x+y; 
+  }
+  
+  EVE_FORCEINLINE auto complex_binary_dispatch( eve::tag::average_
+                                              , auto const& z1, auto const& z2
+                                              ) noexcept
+  {
+    return eve::max ( eve::average(real(z1), real(z2))
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/arithmetic.hpp:73]: why max ? Isn't the average of a complex the average of each of its part ?

File: include/eve/module/complex/regular/detail/arithmetic.hpp [hunk 5]
```diff
@@ -33,21 +42,55 @@ namespace eve::detail
...
+                    );
+  }
 
   template<typename Z>
   EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::dist_, Z const& x, Z const& y) noexcept
   {
     return eve::abs(x-y);
   }
+
+  template<typename Z>
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/arithmetic.hpp:84]: again, not needed

File: include/eve/module/complex/regular/detail/arithmetic.hpp [hunk 6]
```diff
@@ -33,21 +42,55 @@ namespace eve::detail
...
+  EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::div_, Z const& x, Z const& y) noexcept
+  {
+    return x/y;
+  }
+  
+  template<typename Z>
+  EVE_FORCEINLINE auto complex_binary_dispatch(eve::tag::mul_, Z const& x, Z const& y) noexcept
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/arithmetic.hpp:91]: not needed

File: include/eve/module/complex/regular/detail/asin.hpp [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+        }
+        not_treated = logical_notand(zone4, not_treated);
+        if (eve::any(not_treated))
+        {
+          rtype aa = eve::sqrt(eve::inc(eve::sqr(y)));
+          r = if_else(not_treated, x/aa, r);
+          i = if_else(not_treated, half(as(a0r))*eve::log1p(2*y*(y+aa)), i);
+        }
+      }
+      if (eve::any(is_not_finite(a0)))
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/asin.hpp:123]: consider pedantic for this

File: include/eve/module/complex/regular/detail/math.hpp [hunk 8]
```diff
@@ -10,18 +10,247 @@
...
+#include <eve/module/complex/regular/exp_i.hpp>
+#include <eve/module/complex/regular/exp_ipi.hpp>
 
 namespace eve::detail
 {
-  EVE_FORCEINLINE auto complex_binary_dispatch( eve::tag::ulpdist_
-                                              , auto const& z1, auto const& z2
+
+  //==============================================================================================
+  // sqrt cosh cos acosh asinh atan exp exp_i exp_ipi log rec
```
Comment (jfalcou) [inline include/eve/module/complex/regular/detail/math.hpp:22]: sqrt, exp_i and exp_ipi are right there ...

File: include/eve/module/complex/regular/polar.hpp [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+  EVE_MAKE_CALLABLE(polar_, polar);
+
+  namespace detail
+  {
+    template<floating_value V,  floating_value U>
+    EVE_FORCEINLINE auto polar_( EVE_SUPPORTS(cpu_)
+                               , V const & rho
+                               , U const & theta) noexcept
+    {
+      return arithmetic_call(polar, rho, theta);
```
Comment (jfalcou) [inline include/eve/module/complex/regular/polar.hpp:61]: I don't understand why we do an arithmtic_call here

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/complex/regular/exp_i.hpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+  //! **Return value**
+  //!  a complex value
+  //!
+  //! #### Example
+  //!
+  //! @godbolt{doc/complex/exp_i.cpp}
+  //!
+  //!  @}
+  //================================================================================================
+//   namespace tag { struct exp_i_; }
```
Comment (jfalcou) [inline include/eve/module/complex/regular/exp_i.hpp:48]: clean up don't common

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/module/complex/regular/detail/arithmetic.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/complex/regular/detail/arithmetic.hpp:73]: corrected

File: include/eve/module/complex/regular/detail/math.hpp (hunk 8 above)
Comment (jtlap) [reply include/eve/module/complex/regular/detail/math.hpp:22]: forgot the point at the end of the line

File: include/eve/module/complex/regular/exp_i.hpp (hunk 10 above)
Comment (jtlap) [reply include/eve/module/complex/regular/exp_i.hpp:48]: done

File: examples/tutorial/mandelbrot.cpp [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+using wide_i = eve::wide<int>;
+using wide_l = eve::as_logical_t<wide_t>;
+using cwide_t= eve::wide<eve::complex<T>>;
+
+struct mandelbrot
+{
+  static constexpr auto algt = eve::alignment_v<T>;
+  int size, max_iter;
+  T x_min, x_max, y_min, y_max, x_range, y_range;
+  /*eve::detail::alignas(algt)*/ std::vector<int> iterations;
```
Comment (jfalcou) [inline examples/tutorial/mandelbrot.cpp:20]: remove the algt thing there and line 17

**Review by jfalcou — CHANGES_REQUESTED** [review]

**Review by jfalcou — APPROVED** [review]

**Pull request merged**